src/transaction/script.c

```diff
@@ -0,0 +1,257 @@
+#include <stdbool.h>  // bool
+#include <string.h>   // memset, explicit_bzero, memmove
+
+#include "os.h"  // PRINTF
+
+#include "script.h"
+#include "../constants.h"
+#include "opcodes.h"
+
+/**
+ * Identifies if the given script is P2PKH
+ */
+bool identify_p2pkh_script(buffer_t *in, uint16_t script_len) {
+    uint8_t p2pkh[] = {OP_DUP, OP_HASH160, PUBKEY_HASH_LEN, OP_EQUALVERIFY, OP_CHECKSIG};
+    size_t offset = 0;
+
+    if ((in == NULL) || ((script_len != 25) && (script_len != 31)) ||
+        (in->size - in->offset < script_len)) {
+        PRINTF("[-] [Identify P2PKH script] invalid args (script_len=%d, size=%d, offset=%d)\n",
+               script_len,
+               in->size,
+               in->offset);
+        return false;
+    }
+
+    if (script_len == 31) {
+        // check timelock
+        if (in->ptr[in->offset] != 4) {
+            PRINTF("[-] [Identify P2PKH script] invalid timelock: more than 4 bytes\n");
+            return false;
+        }
+        // skip push_4 (1 byte) + timestamp (4 bytes)
+        offset += 5;
+        if (in->ptr[in->offset + offset] != OP_GREATERTHAN_TIMESTAMP) {
+            PRINTF(
+                "[-] [Identify P2PKH script] invalid timelock: OP_GREATERTHAN_TIMESTAMP missing\n");
+            return false;
+        }
+        // skip OP_GREATERTHAN_TIMESTAMP (1 byte)
+        offset += 1;
+    }
+
+    // check OP_DUP, OP_HASH160, push_20
+    if (memcmp(p2pkh, in->ptr + in->offset + offset, 3) != 0) {
+        PRINTF("[-] [Identify P2PKH script] invalid script: starting opcodes mismatch\n");
+        return false;
+    }
+    // Skip opcodes and pubkey hash
+    offset += 3 + PUBKEY_HASH_LEN;
+
+    // check OP_EQUALVERIFY, OP_CHECKSIG
+    if (memcmp(p2pkh + 3, in->ptr + in->offset + offset, 2) != 0) {
+        PRINTF("[-] [Identify P2PKH script] invalid script: ending opcodes mismatch\n");
+        return false;
+    }
+
+    return true;
+}
+
+/**
+ * Identifies if the given script is P2SH
+ */
+bool identify_p2sh_script(buffer_t *in, uint16_t script_len) {
+    uint8_t p2sh[] = {OP_HASH160, PUBKEY_HASH_LEN, OP_EQUAL};
```

Comment:
You are not using the `OP_EQUAL` from here

Reply:
Changed to use the OP_EQUAL from the p2sh array